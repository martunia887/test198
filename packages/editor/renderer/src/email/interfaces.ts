import { Mark } from 'prosemirror-model';

export type NodeSerializer = (opts: NodeSerializerOpts) => string;
export type MarkSerializer = (opts: MarkSerializerOpts) => string;

export type Style = { [key: string]: string | number | undefined };

export interface NodeSerializerOpts {
  attrs: { [key: string]: any };
  marks: Mark[];
  text?: string | null;
}

export interface MarkSerializerOpts {
  mark: Mark;
  text: string;
}

export interface SmartCardWithUrlAttributes {
  url: string;
}

interface ScData {
  '@type': string;
  generator: {
    '@type': string;
    name: string;
  };
  name: string;
  url: string;
  summary: string;
}

export interface SmartCardWithDataAttributes {
  data: ScData;
}
