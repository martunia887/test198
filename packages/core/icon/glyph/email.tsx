/**
 * NOTE:
 *
 * This file is automatically generated by the build process.
 * DO NOT CHANGE IT BY HAND or your changes will be lost.
 *
 * To change the format of this file, modify icon/build/glyph-tsx.js.
 */
import React from "react";
import Icon, { Props } from "../src/components/Icon";

const EmailIcon = (props: Props) => (
  <Icon
    dangerouslySetGlyph={`<svg width="24" height="24" viewBox="0 0 24 24" focusable="false" role="presentation"><g fill="currentColor" fill-rule="evenodd"><path d="M5 7v10h14V7H5zm14-2c1.1 0 2 .9 2 2v10c0 1.1-.9 2-2 2H5c-1.1 0-2-.9-2-2V7c0-1.1.9-2 2-2h14z" fill-rule="nonzero"/><path d="M5.498 6.5H3.124c.149.44.399.854.75 1.205l5.882 5.881a3.117 3.117 0 0 0 4.41 0l5.882-5.881c.35-.351.6-.765.749-1.205h-2.373l-5.672 5.672a1.119 1.119 0 0 1-1.583 0L5.498 6.5z"/></g></svg>`}
    {...props}  
  />
);
EmailIcon.displayName = "EmailIcon";
export default EmailIcon;

