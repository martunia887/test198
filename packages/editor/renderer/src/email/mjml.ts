// import * as mjml2html from 'mjml'
// tslint:disable-next-line:no-var-requires
// const mjml2html = require('mjml').default;
// import mjml2html from 'mjml';
import mjml2html = require('mjml');

import { Fragment, Node as PMNode } from 'prosemirror-model';

// SPIKENOTE:
// MJML is very particular in the way it likes to have it's components nested within each other
// for this reason, it has proven to be very hard to convert ADF structure to mjml.
// It's quite simple to converyt ADF to MJML JSON, but for it to work properly, some
// reshuffling and nesting / wrapping logic needs to be applied.
// Tried this and it seems to be a game of cat+mouse and probably more trouble than what it is worth

// SPIKENOTE: To continue exploration with MJML, we could try to
// create custom mjml components for the differnt ADF nodes.

// import { registerComponent } from 'mjml-core'
// import MjBasicComponent from './mjml-components/MjBasicComponent'
// registerComponent(MjBasicComponent)

// map adf nodes to mjml components
const adfToMjmlTags = new Map([
  ['body', 'mj-body'],
  ['section', 'mj-section'],
  ['text', 'mj-text'],
  // ['heading',   'div'],
  // ['paragraph', 'div'],
  // ['text',      'div'],
  // TODO: add all the things
]);

// fallback node if it is not in the list above
const defaultMjmlNode = 'mj-text';

const adfToMjmlNode = (node: PMNode, children: PMNode[] = []): any => {
  // SPIKENOTE: We could apply some fixes here with conditions?
  // if (adfToMjmlTags.has(<any>node.type)) {

  // }

  const mjmlJson: any = {
    tagName: adfToMjmlTags.get(<any>node.type) || defaultMjmlNode,
    content: node.text,
    attributes: node.attrs,
  };
  if (children.length) {
    mjmlJson.children = children;
  }
  return mjmlJson;
};

export const createMjmlChildren = (node: Fragment): any => {
  let output: any = [];

  node.forEach(childNode => {
    if (!childNode.content) {
      output.push(adfToMjmlNode(childNode));
    } else {
      const convertedChildren = createMjmlChildren(childNode.content);
      output.push(adfToMjmlNode(childNode, convertedChildren));
    }
  });
  return output;
};

export const createMjmlDocument = (adfDocument: Fragment) => ({
  tagName: 'mjml',
  attributes: {},
  children: [
    {
      tagName: 'mj-body',
      attributes: {},
      children: createMjmlChildren(adfDocument),
    },
  ],
});

export const renderMjmlDocument = (mjmlDocument: any): any => {
  return mjml2html(mjmlDocument, {
    beautify: true,
  });
};
