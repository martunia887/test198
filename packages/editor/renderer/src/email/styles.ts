import { fontFamily, fontSize } from '@atlaskit/theme';
import { paragraphStyles } from './nodes/paragraph';
export default `
  .wrapper {
    font-family: ${fontFamily()};
    font-size: ${fontSize()}px;
    font-weight: 400;
    line-height: 24px;
  }
  table {
    font-family: ${fontFamily()};
    font-size: ${fontSize()}px;
    font-weight: 400;
    line-height: 24px;
  }
  .tableNode td > :first-child,
  .tableNode th > :first-child {
    padding-top: 0px;
  }
  .tableNode td > :last-child,
  .tableNode th > :last-child{
    margin-bottom: 0px;
  }
  ${paragraphStyles}
`;
