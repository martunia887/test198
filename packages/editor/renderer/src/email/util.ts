import { Mark } from 'prosemirror-model';
import { Style } from './interfaces';
import { markSerializers } from './serializers';
import { commonStyle } from '.';

export const createTag = (
  tagName: string,
  attrs?: { [key: string]: string | number | undefined },
  content?: string | null,
) => {
  const attrsList: string[] = [];

  Object.keys(attrs || {}).forEach(key => {
    const value = attrs![key];

    if (value === undefined) {
      return;
    }

    attrsList.push(`${key}="${String(value).replace(/"/g, "'")}"`);
  });

  const attrsSerialized = attrsList.length ? ` ${attrsList.join(' ')}` : '';

  return content
    ? `<${tagName}${attrsSerialized}>${content}</${tagName}>`
    : `<${tagName}${attrsSerialized}/>`;
};

export const serializeStyle = (style: Style): string => {
  return Object.keys(style).reduce((memo, key) => {
    if (style[key] === undefined) {
      return memo;
    }

    const value = String(style[key]).replace(/"/g, "'");
    return (memo += `${key}: ${value};`);
  }, '');
};

export const applyMarks = (marks: Mark[], text: string): string => {
  let output = text;
  for (const mark of marks) {
    // ignore marks with unknown type
    if (markSerializers[mark.type.name]) {
      output = markSerializers[mark.type.name]({ mark, text: output });
    }
  }

  return output;
};

type TableData = {
  text?: string | null;
  style: Style;
};

export const commonTableStyle = {
  ...commonStyle,
  margin: '0px',
  padding: '0px',
  'border-spacing': '0px',
  width: '100%',
};

export const createTable = (
  tableData: TableData[][],
  tableStyle: Style = {},
): string => {
  // Tables override font size, weight and other stuff, thus we reset it here with commonStyle
  const attrs = {
    cellspacing: 0,
    cellpadding: 0,
    border: 0,
    style: serializeStyle({
      ...commonTableStyle,
      // Allow overriding any tableStyle, via tableStyle param
      ...tableStyle,
    }),
  };

  const tableRows = tableData.map(tableRow => {
    const tableColumns = tableRow.map(({ style, text }) => {
      const css = serializeStyle(style);
      return createTag('td', { style: css }, text ? text : '');
    });
    return createTag('tr', {}, tableColumns.join(''));
  });

  return createTag('table', attrs, tableRows.join(''));
};
