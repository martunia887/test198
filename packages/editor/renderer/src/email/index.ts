import { fontFamily, fontSize } from '@atlaskit/theme';

import { Fragment, Node as PMNode, Schema } from 'prosemirror-model';

import { Serializer } from '../serializer';
import { nodeSerializers } from './serializers';
import styles from './styles';
import juice from 'juice';
import { escapeHtmlString } from './util';
import flow from 'lodash.flow';
import property from 'lodash.property';
import { defaultSchema as schema } from '@atlaskit/adf-schema';

const serializeNode = (
  node: PMNode,
  index: any,
  parent?: PMNode,
  serializedHTML?: string,
): string => {
  // ignore nodes with unknown type
  if (!nodeSerializers[node.type.name]) {
    return `[UNKNOWN_NODE_TYPE: ${node.type.name}]`;
  }

  const parentAttrs = getAttrsFromParent(index, parent);

  return nodeSerializers[node.type.name]({
    node,
    attrs: {
      ...node.attrs,
      ...parentAttrs,
    },
    marks: node.marks,
    parent: parent,
    text:
      serializedHTML || node.attrs.text || node.attrs.shortName || node.text,
  });
};

/**
 * Used to pass attributes that affect nested nodes.
 *
 * Example: A 'table' node contains 'isNumberColumnEnabled' flag. In order to render
 * numbered columns, 'tableRow' node needs to know this information, thus this function.
 *
 * @param parent {PMNode} parent node
 * @param index {number} index of current child in parent's content array
 */
const getAttrsFromParent = (
  index: number,
  parent?: PMNode,
): { [key: string]: any } => {
  if (parent && parent.attrs && parent.attrs.isNumberColumnEnabled) {
    return {
      index: index,
      isNumberColumnEnabled: true,
    };
  }
  return {};
};

const traverseTree = (fragment: Fragment, parent?: PMNode): string => {
  let output = '';
  fragment.forEach((childNode, _offset, idx) => {
    if (childNode.isLeaf) {
      output += serializeNode(childNode, idx, parent);
    } else {
      const innerHTML = traverseTree(childNode.content, childNode);
      output += serializeNode(childNode, idx, parent, innerHTML);
    }
  });

  return output;
};

export const commonStyle = {
  'font-family': fontFamily(),
  'font-size': `${fontSize()}px`,
  'font-weight': 400,
  'line-height': '24px',
};

const juicify = (html: string): string =>
  juice(`<style>${styles}</style><div class="wrapper">${html}</div>`);

export default class EmailSerializer implements Serializer<string> {
  serializeFragment: (fragment: Fragment) => string = flow(
    (fragment: Fragment) => fragment.toJSON(),
    JSON.stringify,
    escapeHtmlString,
    JSON.parse,
    content => ({ version: 1, type: 'doc', content }),
    schema.nodeFromJSON,
    property('content'),
    traverseTree,
    juicify,
  );

  static fromSchema(_schema: Schema): EmailSerializer {
    return new EmailSerializer();
  }
}
