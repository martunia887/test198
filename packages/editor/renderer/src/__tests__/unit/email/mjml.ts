import { createMjmlDocument, renderMjmlDocument } from '../../../email/mjml';

// WIP : ABANDONED

describe('Renderer - Email - ADF to MJML', () => {
  // tslint:disable-next-line:no-only-tests
  it.only('chould convert adf -> mjml', () => {
    const input: any = {
      type: 'doc',
      version: 1,
      content: [
        {
          type: 'heading',
          content: [
            {
              type: 'text',
              text: 'inlineCard with url',
            },
          ],
          attrs: {
            level: 3,
          },
        },
        {
          type: 'paragraph',
          content: [
            {
              type: 'text',
              text: 'Lorem ipsum',
            },
            {
              type: 'inlineCard',
              attrs: {
                url:
                  'https://extranet.atlassian.com/pages/viewpage.action?pageId=3088533424',
              },
            },
            {
              type: 'text',
              text: 'dolor sit amet',
            },
          ],
        },
      ],
    };
    const mjmlDocument = createMjmlDocument(input.content);
    // console.log(JSON.stringify(mjmlDocument, null, 2))
    expect(mjmlDocument).toMatchSnapshot();
  });

  // tslint:disable-next-line:no-only-tests
  it.only('chould convert very simple adf -> mjml -> html', () => {
    const input: any = {
      type: 'doc',
      version: 1,
      content: [
        {
          type: 'text',
          text: '!!!!!!! -----> SIMPLE TEXT <----- !!!!!! ',
        },
      ],
    };
    const mjmlDocument = createMjmlDocument(input.content);
    const htmlDocument = renderMjmlDocument(mjmlDocument);
    // tslint:disable-next-line:no-console
    console.log(JSON.stringify(mjmlDocument, null, 2));
    // tslint:disable-next-line:no-console
    console.log(htmlDocument);
    expect(htmlDocument).toMatchSnapshot();
  });

  // // tslint:disable-next-line:no-only-tests
  // it.skip('should convert slightly more complex adf -> mjml -> html', () => {

  //   const input: any = {
  //     "type": "doc",
  //     "version": 1,
  //     "content": [
  //       {
  //         "type": "heading",
  //         "content": [
  //           {
  //             "type": "text",
  //             "text": "inlineCard with url"
  //           }
  //         ],
  //         "attrs": {
  //           "level": 3
  //         }
  //       },
  //       {
  //         "type": "paragraph",
  //         "content": [
  //           {
  //             "type": "text",
  //             "text": "Lorem ipsum"
  //           },
  //           {
  //             "type": "inlineCard",
  //             "attrs": {
  //               "url": "https://extranet.atlassian.com/pages/viewpage.action?pageId=3088533424"
  //             }
  //           },
  //           {
  //             "type": "text",
  //             "text": "dolor sit amet"
  //           }
  //         ]
  //       }
  //     ]
  //   }
  //   const mjmlDocument = createMjmlDocument(input.content)
  //   const htmlDocument = renderMjmlDocument(mjmlDocument)
  //   // // tslint:disable-next-line:no-console
  //   // console.log(JSON.stringify(mjmlDocument, null, 2))
  //   // // tslint:disable-next-line:no-console
  //   // console.log(htmlDocument)
  //   expect(htmlDocument).toMatchSnapshot()
  // })
});
