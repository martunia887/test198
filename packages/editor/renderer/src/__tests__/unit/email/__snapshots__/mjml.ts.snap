// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer - Email - ADF to MJML chould convert adf -> mjml 1`] = `
Object {
  "attributes": Object {},
  "children": Array [
    Object {
      "attributes": Object {},
      "children": Array [
        Object {
          "attributes": Object {
            "level": 3,
          },
          "children": Array [
            Object {
              "attributes": undefined,
              "content": "inlineCard with url",
              "tagName": "mj-text",
            },
          ],
          "content": undefined,
          "tagName": "mj-text",
        },
        Object {
          "attributes": undefined,
          "children": Array [
            Object {
              "attributes": undefined,
              "content": "Lorem ipsum",
              "tagName": "mj-text",
            },
            Object {
              "attributes": Object {
                "url": "https://extranet.atlassian.com/pages/viewpage.action?pageId=3088533424",
              },
              "content": undefined,
              "tagName": "mj-text",
            },
            Object {
              "attributes": undefined,
              "content": "dolor sit amet",
              "tagName": "mj-text",
            },
          ],
          "content": undefined,
          "tagName": "mj-text",
        },
      ],
      "tagName": "mj-body",
    },
  ],
  "tagName": "mjml",
}
`;

exports[`Renderer - Email - ADF to MJML chould convert very simple adf -> mjml -> html 1`] = `
Object {
  "errors": Array [
    Object {
      "formattedMessage": "Line undefined of undefined (mj-text) â€” mj-text cannot be used inside mj-body, only inside: mj-attributes, mj-column, mj-hero",
      "line": undefined,
      "message": "mj-text cannot be used inside mj-body, only inside: mj-attributes, mj-column, mj-hero",
      "tagName": "mj-text",
    },
  ],
  "html": "<!doctype html>
<html xmlns=\\"http://www.w3.org/1999/xhtml\\" xmlns:v=\\"urn:schemas-microsoft-com:vml\\" xmlns:o=\\"urn:schemas-microsoft-com:office:office\\">

<head>
  <title>
  </title>
  <!--[if !mso]><!-- -->
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
  <!--<![endif]-->
  <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
  <style type=\\"text/css\\">
    #outlook a {
      padding: 0;
    }

    .ReadMsgBody {
      width: 100%;
    }

    .ExternalClass {
      width: 100%;
    }

    .ExternalClass * {
      line-height: 100%;
    }

    body {
      margin: 0;
      padding: 0;
      -webkit-text-size-adjust: 100%;
      -ms-text-size-adjust: 100%;
    }

    table,
    td {
      border-collapse: collapse;
      mso-table-lspace: 0pt;
      mso-table-rspace: 0pt;
    }

    img {
      border: 0;
      height: auto;
      line-height: 100%;
      outline: none;
      text-decoration: none;
      -ms-interpolation-mode: bicubic;
    }

    p {
      display: block;
      margin: 13px 0;
    }
  </style>
  <!--[if !mso]><!-->
  <style type=\\"text/css\\">
    @media only screen and (max-width:480px) {
      @-ms-viewport {
        width: 320px;
      }

      @viewport {
        width: 320px;
      }
    }
  </style>
  <!--<![endif]-->
  <!--[if mso]>
        <xml>
        <o:OfficeDocumentSettings>
          <o:AllowPNG/>
          <o:PixelsPerInch>96</o:PixelsPerInch>
        </o:OfficeDocumentSettings>
        </xml>
        <![endif]-->
  <!--[if lte mso 11]>
        <style type=\\"text/css\\">
          .outlook-group-fix { width:100% !important; }
        </style>
        <![endif]-->
  <!--[if !mso]><!-->
  <link href=\\"https://fonts.googleapis.com/css?family=Ubuntu:300,400,500,700\\" rel=\\"stylesheet\\" type=\\"text/css\\">
  <style type=\\"text/css\\">
    @import url(https://fonts.googleapis.com/css?family=Ubuntu:300,400,500,700);
  </style>
  <!--<![endif]-->
  <style type=\\"text/css\\">
  </style>
</head>

<body>
  <div style=\\"\\">
    <div style=\\"font-family:Ubuntu, Helvetica, Arial, sans-serif;font-size:13px;line-height:1;text-align:left;color:#000000;\\"> !!!!!!! -----> SIMPLE TEXT <----- !!!!!! </div> </div> </body> </html>",
}
`;
